File: src/carina/ui/assets/js/pages/ops.js
/* Ops console — every platform component in one place: live status, an
   evidence-logged read-only SQL workbench, structured logs, and request
   traces. The laptop stand-in for Backstage + OTel→ClickStack. */

import { api } from "../api.js";
import { esc } from "../theme.js";

const PLANE_ORDER = ["data", "compute", "trust", "product"];

export async function renderOps(root) {
  root.innerHTML = `
    <div class="page-head">
      <p class="eyebrow">Operations plane</p>
      <h1>Ops console</h1>
      <p class="lede">The control plane in one place: each component reports its own live state,
        the SQL workbench is a read-only operator lane (every query lands in the evidence chain),
        and the logs and traces panes show what the platform is doing right now.</p>
    </div>
    <div id="ops-components"><div class="spin" role="status" aria-label="Loading"></div></div>

    <h2 class="ops-h2">SQL workbench <span class="muted ops-note">read-only · row-capped · evidence-logged as <code class="inline">ops.query</code></span></h2>
    <div class="card card-pad">
      <textarea id="sql-in" class="sql-input" rows="3" spellcheck="false"
        placeholder="SELECT date, tension_ratio FROM gold_labour_tension ORDER BY date DESC LIMIT 10"></textarea>
      <div class="ops-row">
        <button class="btn btn-primary" id="sql-run" type="button">Run</button>
        <span class="muted" id="sql-meta"></span>
      </div>
      <div id="sql-out"></div>
    </div>

    <h2 class="ops-h2">Logs <span class="muted ops-note">in-process ring buffer — ClickStack/HyperDX in the cluster profile</span></h2>
    <div class="card">
      <div class="ops-row" style="padding:12px 14px 0;">
        <select id="log-comp"><option value="">all components</option></select>
        <button class="btn btn-ghost" id="log-refresh" type="button">Refresh</button>
      </div>
      <div id="log-out" class="ops-scroll"></div>
    </div>

    <h2 class="ops-h2">Traces <span class="muted ops-note">one trace per API request; spans for authz, semantic compile, lane execution</span></h2>
    <div class="card">
      <div class="ops-row" style="padding:12px 14px 0;">
        <button class="btn btn-ghost" id="trace-refresh" type="button">Refresh</button>
      </div>
      <div id="trace-out" class="ops-scroll"></div>
    </div>`;

  const compEl = root.querySelector("#ops-components");
  try {
    await loadComponents(compEl);
  } catch (err) {
    compEl.innerHTML = `<div class="error-box"><strong>Ops API unavailable.</strong><br/>${esc(err.message)}
      <br/><span class="muted">This console needs a live platform (and, under OIDC, an operator group).</span></div>`;
    return;
  }

  wireSql(root);
  await loadLogs(root);
  root.querySelector("#log-refresh").addEventListener("click", () => loadLogs(root));
  root.querySelector("#log-comp").addEventListener("change", () => loadLogs(root));
  await loadTraces(root);
  root.querySelector("#trace-refresh").addEventListener("click", () => loadTraces(root));
}

async function loadComponents(el) {
  const { components } = await api.opsComponents();
  const byPlane = new Map();
  for (const c of components) {
    if (!byPlane.has(c.plane)) byPlane.set(c.plane, []);
    byPlane.get(c.plane).push(c);
  }
  el.innerHTML = PLANE_ORDER.filter((p) => byPlane.has(p)).map((plane) => `
    <p class="eyebrow" style="margin:18px 0 8px;">${esc(plane)} plane</p>
    <div class="ops-grid">
      ${byPlane.get(plane).map((c) => `
        <div class="card card-pad ops-card">
          <div class="ops-card-head">
            <span class="ops-dot ${esc(c.status)}"></span>
            <strong>${esc(c.name)}</strong>
          </div>
          <dl class="ops-kv">
            ${Object.entries(c.detail || {})
              .filter(([, v]) => v !== null && v !== undefined && v !== "")
              .map(([k, v]) => `<div><dt>${esc(k)}</dt><dd>${esc(String(v))}</dd></div>`).join("")}
          </dl>
        </div>`).join("")}
    </div>`).join("");
}

function wireSql(root) {
  const input = root.querySelector("#sql-in");
  const out = root.querySelector("#sql-out");
  const meta = root.querySelector("#sql-meta");
  root.querySelector("#sql-run").addEventListener("click", async () => {
    const sql = input.value.trim();
    if (!sql) return;
    out.innerHTML = '<div class="spin" role="status" aria-label="Running"></div>';
    meta.textContent = "";
    try {
      const t0 = performance.now();
      const r = await api.opsQuery(sql);
      const ms = Math.round(performance.now() - t0);
      meta.textContent = `${r.row_count} row(s)${r.truncated ? ` (capped at ${r.row_cap})` : ""} · lane ${r.lane} · ${ms} ms`;
      out.innerHTML = r.rows.length ? `
        <div class="ops-scroll"><table class="evidence-table">
          <thead><tr>${r.columns.map((c) => `<th>${esc(c)}</th>`).join("")}</tr></thead>
          <tbody>${r.rows.map((row) => `<tr>${row.map((v) =>
            `<td class="mono">${v === null ? '<span class="muted">∅</span>' : esc(v)}</td>`).join("")}</tr>`).join("")}
          </tbody></table></div>`
        : '<p class="muted" style="padding:10px 0 0;">no rows</p>';
    } catch (err) {
      out.innerHTML = `<div class="error-box">${esc(err.message)}</div>`;
    }
  });
}

async function loadLogs(root) {
  const sel = root.querySelector("#log-comp");
  const { logs, components } = await api.opsLogs(sel.value);
  if (sel.options.length <= 1) {
    for (const c of components) sel.insertAdjacentHTML("beforeend",
      `<option value="${esc(c)}">${esc(c)}</option>`);
  }
  root.querySelector("#log-out").innerHTML = logs.length ? `
    <table class="evidence-table">
      <tbody>${logs.map((l) => `
        <tr>
          <td class="mono" style="white-space:nowrap;">${esc(l.ts.slice(11, 23))}</td>
          <td><span class="ops-level ${esc(l.level)}">${esc(l.level)}</span></td>
          <td class="mono">${esc(l.component)}</td>
          <td>${esc(l.message)}</td>
        </tr>`).join("")}
      </tbody></table>`
    : '<p class="muted" style="padding:12px 14px;">no log records yet — interact with the platform and refresh</p>';
}

async function loadTraces(root) {
  const { traces } = await api.opsTraces();
  root.querySelector("#trace-out").innerHTML = traces.length ? traces.map((t) => {
    const total = Math.max(t.duration_ms || 1, ...t.spans.map((s) => (s.offset_ms || 0) + (s.duration_ms || 0)));
    return `
      <details class="ops-trace">
        <summary>
          <span class="mono">${esc(t.start.slice(11, 23))}</span>
          <strong>${esc(t.name)}</strong>
          <span class="muted">${t.spans.length} span(s) · ${t.duration_ms} ms</span>
        </summary>
        <div class="ops-waterfall">
          ${t.spans.map((s) => `
            <div class="ops-span ${esc(s.status)}">
              <span class="ops-span-name mono">${s.parent_id ? "└ " : ""}${esc(s.name)}</span>
              <span class="ops-span-track">
                <span class="ops-span-bar" style="margin-left:${(100 * (s.offset_ms || 0) / total).toFixed(1)}%;width:${Math.max(1, 100 * (s.duration_ms || 0) / total).toFixed(1)}%;"></span>
              </span>
              <span class="mono muted">${s.duration_ms} ms</span>
            </div>`).join("")}
        </div>
      </details>`;
  }).join("") : '<p class="muted" style="padding:12px 14px;">no traces yet — hit any API endpoint and refresh</p>';
}

File: src/carina/ops.py
"""The operations console backend (ops plane, laptop profile).

One place to see every component of the platform and interrogate it:

  components() — the control-plane inventory: each component's status asked
                 of the component itself (evidence verify, anchor coverage,
                 lane attachment, catalog reachability, auth mode, …)
  run_query()  — the operator SQL lane: read-only by construction (sqlglot
                 whitelist, single statement, row-capped), executed through
                 the lane router, and evidence-logged with the operator's
                 identity. Consumers still never get SQL (ADR-0008); this
                 lane exists for the people who run the platform, and it is
                 audited like everything else.
  logs         — an in-process structured-log ring buffer per component
  traces       — an in-process request tracer (root span per API request,
                 child spans around compile/authz/lane work)

The logs and traces stores are the laptop stand-in for OTel → ClickStack:
same panes in the console, a different backend behind them in the cluster
profile.
"""

from __future__ import annotations

import contextlib
import contextvars
import logging
import os
import time
import uuid
from collections import deque
from datetime import datetime, timezone

import duckdb

from . import anchoring, auth, catalog, evidence, lanes, warehouse
from .contracts import load_products

OPS_GROUPS_ENV = "CARINA_OPS_GROUPS"
DEFAULT_OPS_GROUPS = "platform-team@carina.local"
QUERY_ROW_CAP = 500


def ops_groups() -> list[str]:
    return [g for g in os.environ.get(OPS_GROUPS_ENV, DEFAULT_OPS_GROUPS).split(",") if g]


# ---- structured logs (ring buffer) -----------------------------------------

_LOGS: deque = deque(maxlen=500)


class _RingHandler(logging.Handler):
    def emit(self, record: logging.LogRecord) -> None:
        _LOGS.append({
            "ts": datetime.fromtimestamp(record.created, tz=timezone.utc)
                  .isoformat(timespec="milliseconds"),
            "level": record.levelname,
            "component": record.name,
            "message": record.getMessage(),
        })


def install_log_capture() -> None:
    handler = _RingHandler()
    # uvicorn's loggers don't propagate to root — attach to them directly.
    for name in (None, "uvicorn", "uvicorn.access", "uvicorn.error"):
        logger = logging.getLogger(name)
        if not any(isinstance(h, _RingHandler) for h in logger.handlers):
            logger.addHandler(handler)
    root = logging.getLogger()
    if root.level > logging.INFO or root.level == logging.NOTSET:
        root.setLevel(logging.INFO)


def recent_logs(component: str | None = None, limit: int = 200) -> list[dict]:
    rows = [r for r in _LOGS if not component or r["component"].startswith(component)]
    return rows[-limit:][::-1]


def log_components() -> list[str]:
    return sorted({r["component"] for r in _LOGS})


# ---- traces (in-process, OTel-shaped) ---------------------------------------

_TRACES: deque = deque(maxlen=200)
_current_trace: contextvars.ContextVar = contextvars.ContextVar("carina_trace", default=None)
_current_span: contextvars.ContextVar = contextvars.ContextVar("carina_span", default=None)


@contextlib.contextmanager
def trace(name: str, **attributes):
    """Root span; everything inside lands in one trace."""
    t = {"trace_id": uuid.uuid4().hex[:16], "name": name,
         "start": datetime.now(timezone.utc).isoformat(timespec="milliseconds"),
         "spans": []}
    token = _current_trace.set(t)
    try:
        with span(name, **attributes):
            yield t
    finally:
        t["duration_ms"] = round(sum(
            s["duration_ms"] for s in t["spans"] if s["parent_id"] is None), 2)
        _TRACES.append(t)
        _current_trace.reset(token)


@contextlib.contextmanager
def span(name: str, **attributes):
    """Child span; no-op when there is no active trace."""
    t = _current_trace.get()
    if t is None:
        yield None
        return
    s = {"span_id": uuid.uuid4().hex[:8], "parent_id": _current_span.get(),
         "name": name, "offset_ms": None, "duration_ms": None,
         "attributes": dict(attributes), "status": "ok"}
    trace_start = time.monotonic() if not t["spans"] else None
    if "_t0" not in t:
        t["_t0"] = trace_start or time.monotonic()
    start = time.monotonic()
    s["offset_ms"] = round((start - t["_t0"]) * 1000, 2)
    token = _current_span.set(s["span_id"])
    t["spans"].append(s)
    try:
        yield s
    except Exception:
        s["status"] = "error"
        raise
    finally:
        s["duration_ms"] = round((time.monotonic() - start) * 1000, 2)
        _current_span.reset(token)


def recent_traces(limit: int = 50) -> list[dict]:
    out = []
    for t in list(_TRACES)[-limit:][::-1]:
        out.append({k: v for k, v in t.items() if k != "_t0"})
    return out


# ---- the operator SQL lane ---------------------------------------------------

def validate_readonly(sql: str) -> None:
    """Reject anything that is not a single read statement."""
    import sqlglot
    from sqlglot import expressions as exp

    statements = [s for s in sqlglot.parse(sql, read="duckdb") if s is not None]
    if len(statements) != 1:
        raise ValueError("exactly one statement per query")
    stmt = statements[0]
    if not isinstance(stmt, (exp.Select, exp.Union, exp.Describe, exp.Show)):
        raise ValueError(
            f"read-only lane: {stmt.key.upper()} is not allowed (SELECT/DESCRIBE/SHOW only)")


_log = logging.getLogger("carina.ops")


def run_query(con: duckdb.DuckDBPyConnection, sql: str, actor_subject: str) -> dict:
    validate_readonly(sql)
    _log.info("sql-workbench query by %s: %s", actor_subject,
              sql if len(sql) < 120 else sql[:117] + "...")
    router = lanes.LaneRouter(con)
    decision = router.route([])
    with span("lane.execute", lane=decision.lane):
        res = con.execute(sql)
        columns = [d[0] for d in res.description] if res.description else []
        rows = res.fetchmany(QUERY_ROW_CAP + 1)
    truncated = len(rows) > QUERY_ROW_CAP
    rows = [[None if v is None else str(v) for v in row] for row in rows[:QUERY_ROW_CAP]]
    evidence.record(con, "ops.query", "sql-workbench", {
        "sql": sql, "rows": len(rows), "truncated": truncated,
        "lane": decision.lane, "actor": actor_subject,
    }, actor=actor_subject)
    return {"columns": columns, "rows": rows, "row_count": len(rows),
            "truncated": truncated, "row_cap": QUERY_ROW_CAP,
            "lane": decision.lane}


# ---- component inventory ------------------------------------------------------

def _safe(fn, fallback=None):
    try:
        return fn()
    except Exception as e:
        return fallback if fallback is not None else {"error": f"{type(e).__name__}: {e}"}


def components(con: duckdb.DuckDBPyConnection) -> list[dict]:
    """Every platform component, its plane, and its live status — asked of
    the component itself, not of a config file."""
    out = []

    tables = con.execute(
        "SELECT table_name FROM information_schema.tables WHERE table_schema='main'"
    ).fetchall()
    names = [t[0] for t in tables]
    layers = {
        "bronze": sum(n.startswith("bronze") for n in names),
        "silver": sum(n.startswith("silver") for n in names),
        "gold": sum(n.startswith("gold") for n in names),
    }
    out.append({"id": "lakehouse", "name": "Lakehouse (DuckDB)", "plane": "data",
                "status": "ok" if layers["gold"] else "warn",
                "detail": {**layers, "hint": None if layers["gold"] else "run `carina run`"}})

    ch = evidence.verify(con)
    an = _safe(lambda: anchoring.verify_anchors(con), {"ok": False, "anchors": 0})
    out.append({"id": "evidence", "name": "Evidence chain", "plane": "trust",
                "status": "ok" if ch.get("ok") and an.get("ok") else "error",
                "detail": {"records": ch.get("records", 0), "chain_ok": ch.get("ok"),
                           "anchors": an.get("anchors", 0),
                           "anchored_through": an.get("anchored_through", 0),
                           "unanchored": an.get("unanchored", 0)}})

    router = lanes.LaneRouter(con)
    for lane in router.lanes():
        out.append({"id": f"lane-{lane.id}", "name": f"Lane: {lane.id}", "plane": "compute",
                    "status": "ok" if lane.attached else "off",
                    "detail": {"engine": lane.engine,
                               "attached": lane.attached,
                               "note": lane.description}})

    cat = _safe(catalog.status, {"kind": "unknown", "reachable": False})
    cat_tables = sum(len(v) for v in (cat.get("tables") or {}).values())
    out.append({"id": "catalog", "name": "Catalog (Lakekeeper seam)", "plane": "data",
                "status": "ok" if cat.get("reachable") else
                          ("off" if cat.get("kind") in ("none", "unknown") else "error"),
                "detail": {"kind": cat.get("kind"), "uri": cat.get("uri"),
                           "tables": cat_tables, "error": cat.get("error")}})

    out.append({"id": "warehouse", "name": "Warehouse (Iceberg/Parquet)", "plane": "data",
                "status": "ok" if warehouse.iceberg_available() else "warn",
                "detail": {"format": "iceberg" if warehouse.iceberg_available() else "parquet",
                           "root": str(warehouse.warehouse_root())}})

    out.append({"id": "identity", "name": "Identity (Keycloak seam)", "plane": "trust",
                "status": "ok" if auth.mode() == "oidc" else "warn",
                "detail": {"auth_mode": auth.mode(), "issuer": auth.issuer(),
                           "note": None if auth.mode() == "oidc" else "dev identity — not for shared environments"}})

    denies = con.execute(
        "SELECT count(*) FROM main.evidence WHERE action = 'authz.deny'").fetchone()[0]
    products = load_products()
    n_contracts = sum(len(p.contracts) for p in products)
    out.append({"id": "policy", "name": "Policy (compiled OPA/FGA)", "plane": "trust",
                "status": "ok",
                "detail": {"contracts": n_contracts, "products": len(products),
                           "authz_denies_logged": denies}})

    for p in products:
        pub = evidence.latest(con, "wap.publish", p.id)
        abort = evidence.latest(con, "wap.abort", p.id)
        aborted_last = bool(abort and (not pub or abort["seq"] > pub["seq"]))
        try:
            q = con.execute(
                """SELECT count(*) FILTER (passed), count(*) FROM quality_runs
                   WHERE run_ts = (SELECT max(run_ts) FROM quality_runs q2
                                   WHERE q2.contract = quality_runs.contract)
                     AND contract IN (SELECT unnest(?::VARCHAR[]))""",
                [[ct.id for ct in p.contracts]],
            ).fetchone()
        except duckdb.Error:
            q = (0, 0)  # quality_runs not built yet
        out.append({"id": f"product-{p.id}", "name": f"Product: {p.id}", "plane": "product",
                    "status": "error" if aborted_last else
                              ("ok" if pub and q[0] == q[1] else "warn"),
                    "detail": {"contracts": len(p.contracts),
                               "checks": f"{q[0]}/{q[1]}",
                               "last_publish": (pub or {}).get("ts"),
                               "wap": "ABORTED" if aborted_last else ("published" if pub else "never run")}})

    conf = evidence.latest(con, "conformance.run")
    out.append({"id": "conformance", "name": "Cross-lane conformance", "plane": "compute",
                "status": ("ok" if conf["payload"].get("green") else "error") if conf else "off",
                "detail": {"last_run": (conf or {}).get("ts"),
                           "lanes": (conf or {}).get("payload", {}).get("lanes"),
                           "green": (conf or {}).get("payload", {}).get("green")}})

    return out
